src/backend/distributed/transaction/remote_transaction.c

```diff
@@ -0,0 +1,611 @@
+ */
+void
+MarkRemoteTransactionFailed(MultiConnection *connection, bool allowErrorPromotion)
+{
+	RemoteTransaction *transaction = &connection->remoteTransaction;
+
+	transaction->transactionFailed = true;
+
+	/*
+	 * If the connection is marked as critical, fail the entire coordinated
+	 * transaction. If allowed.
+	 */
+	if (transaction->transactionCritical && allowErrorPromotion)
+	{
+		ereport(ERROR, (errmsg("failure on connection marked as essential: %s:%d",
+							   connection->hostname, connection->port)));
+	}
+}
+
+
+/*
+ * MarkRemoteTransactionCritical signals that failures on this remote
+ * transaction should fail the entire coordinated transaction.
+ */
+void
+MarkRemoteTransactionCritical(struct MultiConnection *connection)
+{
+	RemoteTransaction *transaction = &connection->remoteTransaction;
+
+	Assert(transaction->transactionState != REMOTE_TRANS_INVALID);
+	transaction->transactionCritical = true;
+}
+
+
+/*
+ * CoordinatedRemoteTransactionsCommit performs distributed transactions
+ * handling at commit time. This will be called at XACT_EVENT_PRE_COMMIT if
+ * 1PC commits are used - so shards can still be invalidated - and at
+ * XACT_EVENT_COMMIT if 2PC is being used.
+ *
+ * Note that this routine has to issue rollbacks for failed transactions.
+ */
+void
+CoordinatedRemoteTransactionsCommit(void)
+{
+	dlist_iter iter;
+
+	/*
+	 * Before starting to commit on any of the nodes - after which we can't
+	 * completely roll-back anymore - check that things are in a good state.
+	 */
+	CheckTransactionHealth();
+
+	/*
+	 * Issue appropriate transaction commands to remote nodes. If everything
+	 * went well that's going to be COMMIT or COMMIT PREPARED, if individual
+	 * connections had errors, some or all of them might require a ROLLBACK.
+	 *
+	 * First send the command asynchronously over all connections.
+	 */
+	dlist_foreach(iter, &InProgressTransactions)
+	{
+		MultiConnection *connection = dlist_container(MultiConnection, transactionNode,
+													  iter.cur);
+		RemoteTransaction *transaction = &connection->remoteTransaction;
+
+		Assert(transaction->transactionState != REMOTE_TRANS_INVALID);
+
+		if (transaction->transactionFailed)
+		{
+			transaction->transactionState = REMOTE_TRANS_1PC_ABORTING;
+
+			/*
+			 * Try sending an ROLLBACK; Depending on the state that won't
+			 * succeed, but let's try.  Have to clear previous results
+			 * first.
+			 */
+			ForgetResults(connection); /* try to clear pending stuff */
+			if (!SendRemoteCommand(connection, "ROLLBACK;"))
+			{
+				/* no point in reporting a likely redundant message */
+			}
+		}
+		else if (transaction->transactionState == REMOTE_TRANS_PREPARED)
+		{
+			StringInfoData command;
+
+			initStringInfo(&command);
+			appendStringInfo(&command, "COMMIT PREPARED '%s';",
+							 transaction->preparedName);
+
+			transaction->transactionState = REMOTE_TRANS_2PC_COMMITTING;
+
+			if (!SendRemoteCommand(connection, command.data))
```

Comment:
I didn't do that initially because we should really wait on all the connections in parallel. But I think we can do that by having
```
extern void StartRemoteTransactionCommit(struct MultiConnection *connection);
extern void FinishRemoteTransactionCommit(struct MultiConnection *connection);
extern void RemoteTransactionCommit(struct MultiConnection *connection);
```
where ```StartRemoteTransactionCommit``` and ```FinishRemoteTransactionCommit``` are used by the central handler and ```RemoteTransactionCommit``` just wraps those.  To make that more efficient we later can add a "WaitForIOOnAll(List *connections)" or such.

Will do that.